internals/daemon/api_notices.go

```diff
@@ -0,0 +1,178 @@
+		Types: types,
+		Keys:  keys,
+		After: after,
+	}
+	var notices []*state.Notice
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	timeoutStr := query.Get("timeout")
+	if timeoutStr != "" {
+		// Wait up to timeout for notices matching given filter to occur
+		timeout, err := time.ParseDuration(timeoutStr)
+		if err != nil {
+			return statusBadRequest("invalid timeout %q: %v", timeoutStr, err)
+		}
+
+		ctx, cancel := context.WithTimeout(r.Context(), timeout)
+		defer cancel()
+
+		notices, err = st.WaitNotices(ctx, filter)
+		if errors.Is(err, context.Canceled) {
+			return statusBadRequest("request canceled")
+		}
+		// DeadlineExceeded will occur if timeout elapses; in that case return
+		// an empty list of notices, not an error.
+		if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+			return statusInternalError("cannot wait for notices: %s", err)
+		}
+	} else {
+		// No timeout given, fetch currently-available notices
+		notices = st.Notices(filter)
+	}
+
+	if len(notices) == 0 {
+		notices = []*state.Notice{} // avoid null result
+	}
+	return SyncResponse(notices)
+}
+
+func v1PostNotices(c *Command, r *http.Request, _ *UserState) Response {
+	var payload struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after"`
+		Data        map[string]string `json:"data"`
+	}
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return statusBadRequest("cannot decode request body: %v", err)
+	}
+
+	if payload.Action != "add" {
+		return statusBadRequest("invalid action %q", payload.Action)
+	}
+	if payload.Type != "custom" {
+		return statusBadRequest(`invalid type %q (can only add "custom" notices)`, payload.Type)
+	}
+	if !customKeyRegexp.MatchString(payload.Key) {
+		return statusBadRequest(`invalid key %q (must be in "domain.com/key" format)`, payload.Key)
+	}
+	if len(payload.Key) > maxNoticeKeyLength {
+		return statusBadRequest("key must be %d bytes or less", maxNoticeKeyLength)
+	}
+
+	var repeatAfter time.Duration
+	if payload.RepeatAfter != "" {
+		var err error
+		repeatAfter, err = time.ParseDuration(payload.RepeatAfter)
+		if err != nil {
+			return statusBadRequest("invalid repeat-after duration %q: %v", payload.RepeatAfter, err)
+		}
+	}
+
+	dataSize := 0
+	for k, v := range payload.Data {
+		dataSize += len(k) + len(v)
+	}
+	if dataSize > maxNoticeDataSize {
+		return statusBadRequest("total size of data (keys and values) must be %d bytes or less", maxNoticeDataSize)
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	noticeId, err := st.AddNotice(state.CustomNotice, payload.Key, &state.AddNoticeOptions{
+		Data:        payload.Data,
+		RepeatAfter: repeatAfter,
+	})
+	if err != nil {
+		return statusInternalError("%v", err)
+	}
+
+	result := struct {
+		ID string `json:"id"`
```

Comment:
This is an API contract, so even if small, it probably deserves a proper type so future generations respect it a bit more. Is there a reason for this to not return the real `&Notice{ID: id}`?

Reply:
I considered that, but the only field that's really interesting here is the ID. If we use `&Notice{ID: id}` directly, our `MarshalJSON` would include zero values for the `type`, `key`, `occurrences` and other fields -- which are normally required, so it seems problematic to make them "omitempty".

We could update `State.AddNotice` to return a `*Notice` instead of a `noticeID string`, and then this API would return a full `*Notice`. I guess that would have the nice property that the caller would get the exact times for `first-occurred` and so on, which might be useful?

Reply:
Per voice discussion, we'll add new daemon-local type `addedNotice` or similar with just the ID field.